### PR TITLE
feat(header): replace header user menu with avatar button

### LINKED ISSUE #156: [FEATURE] <Show avatar in Header instead of name>
### Description

This is related #140.

I was thinking that we should go back to showing avatar in the Header instead of name. At least show a default avatar for now.

Showing an avatar is the most common pattern. And also when the size of the name varies too much it makes the design look bad.

Would love to know everyone's opinion on this.

### PR DESCRIPTION
**Pull request by mark-omarov**
### 🛠️ Fixes Issue

Closes #156 

### 👨‍💻 Changes proposed

- Use an avatar-like icon button for the user menu
Since we don't have an actual avatar just yet, in order to keep the design consistent and well-looking I'm proposing to use an avatar icon button similar to the notification bell (see the screenshot below).
Once we get an actual avatar we'll swap it with an actually rounded avatar.
- Extract user menu to standalone `UserMenu` component
- Replace custom className with the `classnames` library
This will allow project-wide usage, instead of custom utilitarian functions
- Extract sign out to a standalone hook
There's still room for improvement, I believe all authentication flow should be encapsulated in the form of hooks.
But, it's outside of the scope of this PR.

### 📷 Screenshots
Screen Shot 2022-09-04 at 12 24 48 AM

### CODE REVIEW

**Review by rohitdasu — APPROVED** [review]

File: src/components/Header/UserMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	to: string;
+};
+
+const itemDefaultClassName = 'block px-4 py-2 text-sm text-gray-700';
+const menuLinks: MenuLink[] = [
+	{ label: 'Profile', to: '#' },
+	{ label: 'Settings', to: '#' },
+];
+
+export function UserMenu({ user, className = '' }: UserMenuPros) {
```
Comment (iShibi) [inline src/components/Header/UserMenu.tsx:25]:
```suggestion
export function UserMenu({ user, className = '' }: UserMenuProps) {
```

File: src/components/Header/UserMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,78 @@
+import { Fragment } from 'react';
+import { HiUser } from 'react-icons/hi';
+import { Menu, Transition } from '@headlessui/react';
+import { type User } from '@supabase/gotrue-js/src/lib/types';
+import cn from 'classnames';
+
+import { useSignOut } from './hooks';
+
+export type UserMenuPros = {
```
Comment (iShibi) [inline src/components/Header/UserMenu.tsx:9]: typo
```suggestion
export type UserMenuProps = {
```
also using interfaces for object types would be better

File: src/components/Header/UserMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
+import { Fragment } from 'react';
+import { HiUser } from 'react-icons/hi';
+import { Menu, Transition } from '@headlessui/react';
+import { type User } from '@supabase/gotrue-js/src/lib/types';
+import cn from 'classnames';
+
+import { useSignOut } from './hooks';
```
Comment (iShibi) [inline src/components/Header/UserMenu.tsx:7]: Remove the empty line
```suggestion
import cn from 'classnames';
import { useSignOut } from './hooks';
```

File: src/components/Header/UserMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	className?: string;
+};
+
+type MenuLink = {
+	label: string;
+	to: string;
+};
+
+const itemDefaultClassName = 'block px-4 py-2 text-sm text-gray-700';
+const menuLinks: MenuLink[] = [
```
Comment (iShibi) [inline src/components/Header/UserMenu.tsx:20]: I suggest we use the generic way for better readability
```suggestion
const menuLinks: Array<MenuLink> = [
```

**Review by iShibi — CHANGES_REQUESTED** [review]

**Comment by iShibi** [comment]
I am unable to see the requirement of creating a `useSignOut` hook. Why not just import the `signOut` function and pass it directly to the `onClick` event?

**Comment by benny123tw** [comment]
I suggest using [`clsx`](https://www.npmjs.com/package/clsx) instead of `classnames` for better performance.

`clsx` serves as a [faster](https://github.com/lukeed/clsx/blob/HEAD/bench) & smaller drop-in replacement for the `classnames` module.

Benchmarks: https://github.com/lukeed/clsx/blob/HEAD/bench

File: src/components/Header/UserMenu.tsx (hunk 4 above)
Comment (benny123tw) [reply src/components/Header/UserMenu.tsx:20]: We could set the rule if you want: https://typescript-eslint.io/rules/array-type/
```ts
"@typescript-eslint/array-type": ["error", { "default": "generic" }]
```

**Comment by mark-omarov** [comment]
> I am unable to see the requirement of creating a `useSignOut` hook. Why not just import the `signOut` function and pass it directly to the `onClick` event?

If you take a look at the existing code (before the change) the function for signing out exists in the space of the Header, if we're taking about this specific function, exporting from the Header, and importing in the UserMenu component would create a circular dependency which is a bad thing. On the other hand, I could have used the sign out functionality from the supabase library, but we have a wrapped functionality that handles the error case, even though this is a very basic error handling, I believe it's crucial to have an encapsulated logic of the authentication with a single source. This is aligned with the DRY principle, and SOLID principle.
I could create a file with the function export, but it would not be very much aligned with how react works, this hook can be moved to a somewhat more appropriate place, renamed to `useAuth`, and enhanced with all authentication functionality. But, as I've mentioned, it's not in the scope of this PR.

**Comment by mark-omarov** [comment]
> I suggest using [`clsx`](https://www.npmjs.com/package/clsx) instead of `classnames` for better performance.
> 
> 
> 
> `clsx` serves as a [faster](https://github.com/lukeed/clsx/blob/HEAD/bench) & smaller drop-in replacement for the `classnames` module.
> 
> 
> 
> Benchmarks: https://github.com/lukeed/clsx/blob/HEAD/bench

I've never seen this before, but I love this, I'll update the PR to use it, thanks for sharing! Please do not merge until I update the PR (sometime tomorrow probably)

**Comment by iShibi** [comment]
> If you take a look at the existing code (before the change) the function for signing out exists in the space of the Header, if we're taking about this specific function, exporting from the Header, and importing in the UserMenu component would create a circular dependency which is a bad thing. On the other hand, I could have used the sign out functionality from the supabase library, but we have a wrapped functionality that handles the error case, even though this is a very basic error handling, I believe it's crucial to have an encapsulated logic of the authentication with a single source. This is aligned with the DRY principle, and SOLID principle.

I was aware of the circular dependency and was referring to putting the function in a new file.

> I could create a file with the function export, but it would not be very much aligned with how react works, this hook can be moved to a somewhat more appropriate place, renamed to `useAuth`, and enhanced with all authentication functionality. But, as I've mentioned, it's not in the scope of this PR.

In react we do create custom hooks to abstract the logic but only when it's needed. In this case, our `signOut` function has no need to be a hook as it is just wrapping error handling. It can easily be a helper function that we can import and pass where needed. Wrapping it in a custom hook is going to hurt the performance with no gain.

I would suggest moving the `signOut` function into `src/utils/auth.ts` and deleting the `hooks` folder.

File: src/components/Header/UserMenu.tsx (hunk 4 above)
Comment (iShibi) [reply src/components/Header/UserMenu.tsx:20]: @benny123tw that would be a nice addition. Can you create a PR for it?

Comment (benny123tw) [reply src/components/Header/UserMenu.tsx:20]: @iShibi sure. I'll create a PR.

File: src/components/Header/UserMenu.tsx (hunk 3 above)
Comment (mark-omarov) [reply src/components/Header/UserMenu.tsx:7]: I prefer a clear distinction between libraries and local file imports, but whatever, not a big deal.

File: src/components/Header/UserMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import { type User } from '@supabase/gotrue-js/src/lib/types';
+import clsx from 'clsx';
+import { signOut } from '../../utils/auth';
+
+export interface UserMenuProps {
+	user: User;
+	className?: string;
+}
+
+type MenuLink = {
```
Comment (iShibi) [inline src/components/Header/UserMenu.tsx:13]:
```suggestion
interface MenuLink {
```

**Review by iShibi — CHANGES_REQUESTED** [review]
Thank you for implementing the suggestions. Just two things and we are good to go.

The email is overflowing in desktop mode. I think we can use some ellipsis here.
See: https://tailwindcss.com/docs/text-overflow#ellipsis

overflow

**Comment by mark-omarov** [comment]
Did not change any of the items styling, but a good catch, probably a good idea to fix it in this PR as well. Thanks!

**Review by benny123tw — APPROVED** [review]

File: src/components/Header/index.ts [hunk 6]
```diff
@@ -0,0 +1 @@
+export { Header } from './Header';
```
Comment (subhoghoshX) [inline src/components/Header/index.ts:1]: I don't think there was any need to create separate components. Especially as the code isn't that big.

Comment (mark-omarov) [reply src/components/Header/index.ts:1]: Header had 242L of code, I don't believe having large components is a good practice. Composition is a much better solution IMO. But you're free to disagree.

**Comment by mark-omarov** [comment]
@iShibi here's what it will look like.
Also, the email will always be on a new line now, figured the shorted ones are collapsed to a single line, didn't look very nice.
Screen Shot 2022-09-05 at 3 35 23 PM

File: src/components/Header/Header.tsx [hunk 7]
```diff
@@ -1,22 +1,13 @@
...
-function classNames(...classes: Array<string>) {
-	return classes.filter(Boolean).join(' ');
-}
+import { HiBell, HiMenu, HiX } from 'react-icons/hi';
+import { Disclosure } from '@headlessui/react';
+import type { UserProps } from '../../lib/types';
+import { signOut } from '../../utils/auth';
+import { UserMenu } from './UserMenu';
 
 export function Header({ user }: UserProps) {
```
Comment (iShibi) [inline src/components/Header/Header.tsx:8]: Can you create a new `HeaderProps` interface for this here? `UserProps` has been renamed to `PageProps` in #161 and will have more page-specific props.

**Review by iShibi — CHANGES_REQUESTED** [review]
@mark-omarov that looks good for now. Once we replace email with username, we can go back to the same line with ellipsis just like `GitHub`.

File: src/components/Header/Header.tsx (hunk 7 above)
Comment (mark-omarov) [reply src/components/Header/Header.tsx]: @iShibi done, my friend, please check it out, I hope I understood your ask correctly.

**Review by iShibi — APPROVED** [review]

**Comment by subhoghoshX** [comment]
Thanks for working on it @mark-omarov. I was a little busy the last few days so couldn't review the PR.

I'm not quite sure about the avatar icon though. Maybe we can change it later. Good work :+1:

**Pull request merged**